eq_cir_converter_service/utils/helper_utils.py

```diff
@@ -0,0 +1,181 @@
+    :return: A list of cleaned paragraphs or dictionaries with placeholders.
+    """
+    raw_text = text_obj.get("text", "")
+    placeholders = text_obj.get("placeholders", [])
+    paragraphs = extract_paragraphs(str(raw_text))
+
+    result: list[str | dict[str, str | list | object]] = []
+    for para in paragraphs:
+        cleaned = clean_html_tags(para).strip()
+        if not cleaned:
+            continue
+
+        used_counts = Counter(extract_placeholders(cleaned))
+        relevant: list[dict] = []
+        for ph_name, count in used_counts.items():
+            if isinstance(placeholders, list):
+                matching = []
+                for placeholder in placeholders:
+                    if isinstance(placeholder, dict) and placeholder.get("placeholder") == ph_name:
+                        # Ensure the placeholder is a dictionary with a 'placeholder' key
+                        matching = [placeholder]
+
+                # If the placeholder is found, create a deep copy for each occurrence
+                if len(matching) > 0:
+                    relevant.extend(deepcopy(matching[0]) for _ in range(count))
+
+        if relevant:
+            result.append({"text": cleaned, "placeholders": relevant})
+        else:
+            result.append(cleaned)
+
+    return result
+
+
+# --- Helper Functions for process_element ---
+def process_string(text: str) -> str | list[str]:
+    """Processes a string, cleaning HTML tags and splitting into paragraphs if necessary.
+
+    :param text: The input string to process.
+    :return: A cleaned string or a list of paragraphs.
+    """
+    if REGEX_PARA_SPLIT.search(text):
+        # If the text contains <p> tags, split into paragraphs
+        # and clean each paragraph
+        paragraphs = [clean_html_tags(p).strip() for p in extract_paragraphs(text) if clean_html_tags(p).strip()]
+        if len(paragraphs) == 1:
+            return paragraphs[0]  # return string if only one paragraph
+        return paragraphs
+    return clean_html_tags(text).strip()
+
+
+def process_text_object(
+    obj: dict[str, str | list | object],
+) -> dict[str, str | list | object] | list[str | dict[str, str | list | object]]:
+    """Processes a text object, cleaning HTML tags and extracting paragraphs with placeholders.
+
+    :param obj: A dictionary containing 'text' and possibly 'placeholders'.
+    :return: A cleaned text object or a list of paragraphs with placeholders.
+    """
+    if REGEX_PARA_SPLIT.search(str(obj.get("text", ""))):
+        # If the text contains <p> tags, split into paragraphs
+        # and clean each paragraph, extracting placeholders
+        return split_text_with_placeholders(obj)
+    obj["text"] = clean_html_tags(str(obj["text"])).strip()
+    return obj
+
+
+def process_list(elements: Sequence[str | Sequence[Any] | dict]) -> list[str | list | object]:
+    """Processes a list of elements, cleaning HTML tags and extracting paragraphs from text objects.
```

Comment:
This description indicates that we work on text objects but wouldn't that description fit the other function `process_text_object`?  Is it repetition of what the `process_text_object` does.